src/IssueViz.Security/Taint/TaintStore.cs

```diff
@@ -62,23 +132,60 @@ public void Set(IEnumerable<IAnalysisIssueVisualization> issueVisualizations, An
                 throw new ArgumentNullException(nameof(issueVisualizations));
             }
 
-            this.analysisInformation = analysisInformation;
+            lock (Locker)
+            {
+                this.analysisInformation = analysisInformation;
+
+                var oldIssues = taintVulnerabilities;
+                taintVulnerabilities = issueVisualizations.ToList();
 
-            var oldIssues = taintVulnerabilities;
-            taintVulnerabilities = issueVisualizations.ToArray();
+                var removedIssues = oldIssues.Except(taintVulnerabilities, TaintAnalysisIssueVisualizationByIssueKeyEqualityComparer.Instance).ToArray();
+                var addedIssues = taintVulnerabilities.Except(oldIssues, TaintAnalysisIssueVisualizationByIssueKeyEqualityComparer.Instance).ToArray();
 
-            NotifyIssuesChanged(oldIssues);
+                NotifyIssuesChanged(removedIssues, addedIssues);
+            }
         }
 
-        private void NotifyIssuesChanged(IAnalysisIssueVisualization[] oldIssues)
+        private void NotifyIssuesChanged(
+            IReadOnlyCollection<IAnalysisIssueVisualization> removedIssues, 
+            IReadOnlyCollection<IAnalysisIssueVisualization> addedIssues)
         {
-            var removedIssues = oldIssues.Except(taintVulnerabilities).ToArray();
-            var addedIssues = taintVulnerabilities.Except(oldIssues).ToArray();
-
             if (removedIssues.Any() || addedIssues.Any())
             {
                 IssuesChanged?.Invoke(this, new IssuesChangedEventArgs(removedIssues, addedIssues));
             }
         }
+
+        private class TaintAnalysisIssueVisualizationByIssueKeyEqualityComparer : IEqualityComparer<IAnalysisIssueVisualization>
```

Comment:
They're not going to be the same instance, so it's really only the `IssueKey` that matters. Why bother with a converter? We do the casting inside FindIndex for Remove, so why not do the same in Add?